# TALLER Clase 25 - Modelos lineales generalizados

Alejandro Mantilla Redondo - 201711304

Ximena Palacio Argüello - 201730995

En el paquete ISLR de R, la base de datos `OJ` contiene información sobre la venta de jugos de naranja (citrus hill `CH` o Minute Maid `MM`), en función de varias variables. Use las variables: precio de CH, precio de MM, descuento en CH, desceunto en MM, y las dos indicadoras de si son productos especiales.

In [1]:
library(ISLR)
datos <- ISLR::OJ[, c("Purchase",
                      "PriceCH",
                      "PriceMM",
                      "DiscCH",
                      "DiscMM",
                      "SpecialCH",
                      "SpecialMM")]

datos$Purchase <- ifelse(datos$Purchase == "CH", 1, 0)

head(datos)


Purchase PriceCH PriceMM DiscCH DiscMM SpecialCH SpecialMM
1 1        1.75    1.99    0.00   0.0    0         0        
2 1        1.75    1.99    0.00   0.3    0         1        
3 1        1.86    2.09    0.17   0.0    0         0        
4 0        1.69    1.69    0.00   0.0    0         0        
5 1        1.69    1.69    0.00   0.0    0         0        
6 1        1.69    1.99    0.00   0.0    0         1

## 1. Estime el modelo.

$$
\ln\left(\frac{p}{1-p}\right) = \beta_0 + \beta_{\text{priceCH}}x_{\text{priceCH}} + \beta_{\text{priceMM}}x_{\text{priceMM}} + \beta_{\text{discCH}}x_{\text{discCH}} + \beta_{\text{discMM}}x_{\text{discMM}}
$$

Se toma como una evento positivo la compra de `CH`. $p$ es por lo tanto la probabilidad estimada de que un registro resulte en la compra de `CH`.

In [2]:
modelo <- glm(Purchase ~.,
              data = datos, family = 'binomial')
summary(modelo)



Call:
glm(formula = Purchase ~ ., family = "binomial", data = datos)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.2208  -1.1650   0.7689   0.9473   1.7500  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  -1.7940     1.2818  -1.400  0.16165    
PriceCH      -2.4158     0.8602  -2.809  0.00498 ** 
PriceMM       3.2613     0.6220   5.243 1.58e-07 ***
DiscCH        3.5402     0.8306   4.262 2.03e-05 ***
DiscMM       -1.2438     0.3816  -3.259  0.00112 ** 
SpecialCH     0.4081     0.2509   1.627  0.10380    
SpecialMM    -0.4951     0.2094  -2.364  0.01808 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1430.9  on 1069  degrees of freedom
Residual deviance: 1318.8  on 1063  degrees of freedom
AIC: 1332.8

Number of Fisher Scoring iterations: 4


## 2. Interprete los parámetros.

### Interceptos

In [3]:
betas <- modelo$coefficients

p = exp(betas["(Intercept)"]) / (1 + exp(betas["(Intercept)"]))
p

(Intercept) 
  0.1425822

El intercepto del modelo nos dice que el valor esperado del logaritmo del _odds ratio_ cuando no es una venta especial (tanto para `CH` como para `MM`), el precio de venta es cero y no hay descuentos, es $-1.7940$ . En terminos de la probabilidad de que la venta sea de `CH`, $p = 0.143$. Esta probabilidad es menor a la que tendríamos en un modelo sin intercepto.

In [4]:
p = (exp(betas["(Intercept)"] + betas["SpecialCH"])
     / (1 + exp(betas["(Intercept)"] + betas["SpecialCH"])))
p

(Intercept) 
  0.2000659

El intercepto del modelo nos dice que el valor esperado del logaritmo del _odds ratio_ cuando no es una venta especial (solo para `MM`), el precio de venta es cero y no hay descuentos, es $-1.386$ . En terminos de la probabilidad de que la venta sea de `CH`, $p = 0.200$. Esta probabilidad es menor a la que tendríamos en un modelo sin intercepto.

In [5]:
p = (exp(betas["(Intercept)"] + betas["SpecialMM"])
     / (1 + exp(betas["(Intercept)"] + betas["SpecialMM"])))
p

(Intercept) 
 0.09202636

El intercepto del modelo nos dice que el valor esperado del logaritmo del _odds ratio_ cuando no es una venta especial (solo para `CH`), el precio de venta es cero y no hay descuentos, es $-2.289$ . En terminos de la probabilidad de que la venta sea de `CH`, $p = 0.092$. Esta probabilidad es menor a la que tendríamos en un modelo sin intercepto.

In [6]:
p = (exp(betas["(Intercept)"]
         + betas["SpecialMM"]
         + betas["SpecialCH"])
     / (1 + exp(betas["(Intercept)"]
                + betas["SpecialMM"]
                + betas["SpecialCH"])))
p

(Intercept) 
  0.1322722

El intercepto del modelo nos dice que el valor esperado del logaritmo del _odds ratio_ cuando es una venta especial (tanto para `CH` como para `MM`), el precio de venta es cero y no hay descuentos, es $-1.881$. En terminos de la probabilidad de que la venta sea de `CH`, $p = 0.132$. Esta probabilidad es menor a la que tendríamos en un modelo sin intercepto.

### Efectos

In [7]:
exp(betas["PriceCH"])

PriceCH 
0.08929727

Si el precio de `CH` incrementa en una unidad, el logaritmo del _odds ratio_ incrementa $\beta_{\text{precioCH}}$ unidades, es decir, $-2.416$. En términos del _odds ratio_, este cambia en un factor de $e^{\beta_{\text{precioCH}}} = 0.089$. El _odds ratio_ disminuye con incrementos en el precio de `CH`.

In [8]:
exp(betas["PriceMM"])

PriceMM 
26.08288

Si el precio de `MM` incrementa en una unidad, el logaritmo del _odds ratio_ incrementa $\beta_{\text{precioMM}}$ unidades, es decir, $3.2613$. En términos del _odds ratio_, este cambia en un factor de $e^{\beta_{\text{precioMM}}} = 26.082$. El _odds ratio_ incrementa con incrementos en el precio de `MM`.

In [9]:
exp(betas["DiscCH"])

DiscCH 
34.47444

Si el descuento de `CH` incrementa en una unidad, el logaritmo del _odds ratio_ incrementa $\beta_{\text{discCH}}$ unidades, es decir, $3.5402$. En términos del _odds ratio_, este cambia en un factor de $e^{\beta_{\text{discCH}}} = 34.474$. El _odds ratio_ incrementa con incrementos en el descuento de `CH`.

In [10]:
exp(betas["DiscMM"])

DiscMM 
0.2882987

Si el descuento de `MM` incrementa en una unidad, el logaritmo del _odds ratio_ incrementa $\beta_{\text{discMM}}$ unidades, es decir, $-1.244$. En términos del _odds ratio_, este cambia en un factor de $e^{\beta_{\text{discMM}}} = 0.288$. El _odds ratio_ disminuye con incrementos en el descuento de `MM`.

## 3. ¿Es el modelo significativo?

### LRT

Relizamos una prueba de significancia global.

\begin{align*}
    H_{0} &: \text{El modelo no es significativo.}\\
    H_{a} &: \neg H_{0}.
\end{align*}

#### Forma corta

In [12]:
ep <- summary(modelo)$deviance - summary(modelo)$null.deviance
pchisq(ep, df = 7 - 1)

[1] 0

In [13]:
ep

[1] -112.0925

#### Forma larga

In [ ]:
comp <- 2 * (sum(datos$Purchase
                * log(modelo$fitted.values))
            + sum((1 - datos$Purchase)
                  * log(1 - modelo$fitted.values)))

red <- 2 * (sum(datos$Purchase
                 * log(mean(datos$Purchase)))
             + sum((1 - datos$Purchase)
                   * log(1 - mean(datos$Purchase))))

In [ ]:
ep <- comp - red
1 - pchisq(ep, df = 7 - 1)

In [15]:
modelo_red <- glm(Purchase ~ 1, data = datos, family = "binomial")

# summary(modelo_red)
# anova(modelo, modelo_red, test = "LRT")

In [17]:
ep <- summary(modelo)$deviance - summary(modelo_red)$deviance
ep

[1] -112.0925

Dado que el $p$-value es menor a $\alpha = 0.05$, rechazamos que el modelo no sea significativo.

## 4. ¿Es cierto que las variables de si el producto es especial no afectan las ventas?

In [ ]:
modelo_red <- glm(Purchase ~ . - SpecialCH - SpecialMM, data = datos, family = "binomial")

ep <- summary(modelo)$null.deviance - summary(modelo_red)$null.deviance
1 - pchisq(ep, df = nrow(datos) - 7)

In [ ]:
summary(modelo_red)

In [ ]:
ep

In [ ]:
modelo_red <- glm(Purchase ~ SpecialCH + SpecialMM, data = datos, family = "binomial")

anova(modelo_red, modelo, test = "LRT")

## 5. ¿Es cierto que el efecto del precio de CH es el negativo del precio de MM?

## 6. Haga un intervalo para la probabilidad de compra de MM si precio CH=1.5, precio MM=2.1, descuento CH=0, descuento MM=.2, Special CH=SpecialMM=0. 

In [ ]:
"
Código clase 23.

En el paquete ISLR de R, la base de datos Default contien información sobre
si las personas que piden un préstamo no pagan (hacen default) o si lo pagan
completo. La variable de respuesta es 'default' y se usan las variables 
explicativa que dicen si la persona es estudiante, el balance de la cuenta
y el ingreso.

1. Estime el modelo
2. Interprete los parámetros
3. Es el modelo significativo?
4. Es cierto que las variables de estudiante y el ingreso se pueden omitir
   en el modelo?
5. Es cierto que el intercepto debe ser menor a 2?
6. Es cierto que beta_2 es 100 veces beta 3?
7. Haga un intervalo para la probabilidad de default para un estudiante
   con balance 900 e ingreso 20000. 
8. De acuerdo al banco, para un estudiante con balance 1000 e ingreso 18000
   la porbabilidad de default es 0.8, mientas para un estudiante con balance
   de 1500 e ingreso de 22000, la porbabilidad es de 0.2. Encuentra evidencia
   para refutar esta información?
"
########################################################
#Solution

library(ISLR)
data=Default

fit=glm(default~.,data=data,family="binomial")
summary(fit)

LR=2920.6-1571.5
pvalor=1-pchisq(LR,3)

pvalor

nul=glm(default~1,data=data,family="binomial")
summary(nul)

anova(nul,fit,test="LRT")

nul2=glm(default~balance,data=data,family="binomial")

anova(nul2,fit,test="LRT")

student2=ifelse(data$student=="Yes",1,0)
data$student=student2

x=as.matrix(cbind(rep(1,dim(data)[1]),data[,2:4]))
head(x)

var_beta=solve(t(x)%*%(diag(fit$fitted.values*(1-fit$fitted.values)))%*%x)

ep=(fit$coefficients[1]-2)/sqrt(var_beta[1,1])

########################################
c=c(0,0,1,-100)
theta=t(fit$coefficients)%*%c

var_theta=t(c)%*%var_beta%*%c

ep2=theta/sqrt(var_theta)


############################################
c=c(1,0,900,20000)
mu=t(c)%*%fit$coefficients

sup=mu+qnorm(0.975)*sqrt(t(c)%*%var_beta%*%c)
inf=mu-qnorm(0.975)*sqrt(t(c)%*%var_beta%*%c)

supp=exp(sup)/(1+exp(sup))
infpp=exp(inf)/(1+exp(inf))
interp=c(infpp,supp)